### PR TITLE
[jvm-packages] automatically set the max/min direction for best score

### PR DESCRIPTION
**Pull request by wbo4958**
To fix https://github.com/dmlc/xgboost/issues/9377

### CODE REVIEW

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java [hunk 1]
```diff
@@ -360,16 +376,50 @@ static boolean shouldEarlyStop(int earlyStoppingRounds, int iter, int bestIterat
...
       String maximize = String.valueOf(params.get("maximize_evaluation_metrics"));
-      assert(maximize != null);
       return Boolean.valueOf(maximize);
-    } catch (Exception ex) {
-      logger.error("maximize_evaluation_metrics has to be specified for enabling early stop," +
-              " allowed value: true/false", ex);
-      throw ex;
     }
+
+    if (params.get("eval_metric") != null) {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:405]: If the user has specified metric name, it should reflect in the log, could you please help confirm that?

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java [hunk 2]
```diff
@@ -360,16 +376,50 @@ static boolean shouldEarlyStop(int earlyStoppingRounds, int iter, int bestIterat
...
+      // user has special metric name
+      metricName = String.valueOf(params.get("eval_metric"));
+    } else {
+      // infer the metric name from log
+      metricName = getMetricNameFromlog(evalInfo, evalNames);
+    }
+
+    assert metricName != null;
+
+    if (!"mape".equals(metricName)) {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:415]: why mape?

**Review by trivialfis — COMMENTED** [review]
Hey, thank you for the fix!

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 1 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:405]: Yes, if the user sets the eval_matric, the log will contain it.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 2 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:415]: I just followed the python way to handle the direction, https://github.com/dmlc/xgboost/blob/f7f673b00c15458fb4dd74a2a0d2ba80369c5faf/python-package/xgboost/callback.py#L384

Comment (trivialfis) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:415]: Ah, that's the `mape` starting with `map`.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 1 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:405]: Do we still need to check whether `eval_metric` is null given that we can obtain all the info from the log?

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 2 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:415]: yes,

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java (hunk 1 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:405]: good point. But let's just keep it there. no need to infer it again.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**